### PR TITLE
feat: Shell `if-else`, conditional expressions, running commands in background

### PR DESCRIPTION
**Pull request by zackradisic**
### What does this PR do?

This adds 3 new features to the shell:
- [if-else](https://pubs.opengroup.org/onlinepubs/009604499/utilities/xcu_chap02.html#tag_02_09_04_07)  (e.g. `if cond; then stuff; else other stuff; fi`)
- [conditional expressions](https://www.gnu.org/software/bash/manual/html_node/Bash-Conditional-Expressions.html) (e.g. `[[ -f package.json ]] && echo package.json exists!`)
- Running [commands in the background](https://pubs.opengroup.org/onlinepubs/009604499/utilities/xcu_chap02.html#tag_02_09_03_02) (e.g. `longrunningcmd &`)

Note that technically commands run in the background should be a separate process with their own PID, but this is not implemented yet

Also, not all conditional expression operators are supported ([here](https://www.gnu.org/software/bash/manual/html_node/Bash-Conditional-Expressions.html) is a list of all them in bash). For an initial implementation, the following are suppported:
- `-d` is directory
- `-f` file exists
- `-c` is character device
- `-z` string is empty
- `-n` string is non-empty

### How did you verify your code works?

- [x] I checked the lifetime of memory allocated to verify it's (1) freed and (2) only freed when it should be
- [x] I included a test for the new code, or an existing test covers it
- [x] JSValue used outside outside of the stack is either wrapped in a JSC.Strong or is JSValueProtect'ed
- [x] I wrote TypeScript/JavaScript tests and they pass locally (`bun-debug test test-file-name.test`)

### CODE REVIEW

File: src/shell/shell.zig
Comment (dylan-conway) [inline src/shell/shell.zig:402]: is `-a` normally a flag? it's already covered by `-e`

Comment (zackradisic) [reply src/shell/shell.zig:402]: Yep, at first I thought it was gpt hallucinating when it generated this struct, but for some reason -a and -e do the same thing according to the gnu bash manual: https://www.gnu.org/software/bash/manual/bash.html#Bash-Conditional-Expressions

Comment (dylan-conway) [reply src/shell/shell.zig:402]: oh interesting

File: src/bun.js/event_loop.zig
Comment (Jarred-Sumner) [inline src/bun.js/event_loop.zig]: Please remove this comment

File: src/bun.js/event_loop.zig
Comment (Jarred-Sumner) [inline src/bun.js/event_loop.zig]: ✂️

File: src/shell/interpreter.zig
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig]: This is not a helpful error message

File: src/shell/interpreter.zig
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig]: Doesn’t this leak memory? Don’t you need to call .destroy?

File: src/shell/interpreter.zig
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig:3363]: This seems very sketchy

Comment (Jarred-Sumner) [reply src/shell/interpreter.zig:3363]: Can you leave a detailed comment explaining why this makes sense?

File: src/shell/interpreter.zig
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig:3515]: Why are we doing special checks for ISCHR? A file, pipe, a socket can all be a TTY and not a character device. You most likely want isatty instead.

File: src/shell/interpreter.zig
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig]:
```suggestion
                    .done => @panic(“This code should not be reachable”) 
```

File: src/shell/shell.zig
Comment (Jarred-Sumner) [inline src/shell/shell.zig]: strings.eql
```suggestion
            strings.eql(self.text(self.peek().Text), @tagName(toktag)))
```

File: src/shell/shell.zig
Comment (Jarred-Sumner) [inline src/shell/shell.zig]: What are these comments for? Should you delete this?

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
Some comments

Definitely not enough tests. We cannot ship this until there are a lot more tests.

**Comment by Jarred-Sumner** [comment]
I’m also confused where the code for subprocesses or job control is. It’d probably be reasonable to ship if else separately from &

**Comment by Jarred-Sumner** [comment]
@zackradisic there are merge conflicts

File: src/shell/interpreter.zig
Comment (zackradisic) [reply src/shell/interpreter.zig:3515]: This is for -c conditional expression which checks that the file is a character device

**Comment by Jarred-Sumner** [comment]
Almost ready, but there are test failures that need to be addressed before this PR can be merged.

Tests are failing on Windows:

- test\cli\install\bun-run-bunfig.test.ts 12 failing
- test\cli\install\bun-run.test.ts 4 failing
- test\cli\run\env.test.ts 10 failing
- test\cli\run\if-present.test.ts 1 failing
- test\cli\run\run-process-env.test.ts 2 failing
- test\cli\run\run-quote.test.ts 1 failing
- test\js\bun\shell\bunshell.test.ts 1 failing 

Tests are failing on macOS/Linux:

- [test/cli/install/bun-run-bunfig.test.ts](https://github.com/oven-sh/bun/pull/9631#testcliinstallbun-run-bunfigtestts) 4 failing
- [test/cli/run/env.test.ts](https://github.com/oven-sh/bun/pull/9631#testclirunenvtestts) 7 failing

**Comment by mrousavy** [comment]
Hey - should background commands work already after this PR was merged? When I run a command like this:

```js
await $`"$ANDROID_HOME/emulator/emulator" -avd "${emulatorName}" -netdelay none -netspeed full &`.quiet()
```

I get this bun error:

```diff
- error: Background commands "&" are not supported yet.
-       at BunShell (191:35)
```

File: src/shell/shell.zig
Comment (mrousavy) [inline src/shell/shell.zig:1027]: Ah, I guess the error is expected then. Hm.

**Pull request merged**
